src/python/pants/task/simple_codegen_task.py

```diff
@@ -214,17 +225,27 @@ def execute(self):
 
       with self.context.new_workunit(name='execute', labels=[WorkUnitLabel.MULTITOOL]):
         for vt in invalidation_check.all_vts:
+
+          sources = None
+          must_recapture = False
+
           # Build the target and handle duplicate sources.
           if not vt.valid:
             if self._do_validate_sources_present(vt.target):
               self.execute_codegen(vt.target, vt.results_dir)
-              self._handle_duplicate_sources(vt.target, vt.results_dir)
+              sources = self._capture_sources(vt.target, vt.results_dir)
```

Comment:
This would be a very good usecase for using the batchness of the `capture_snapshots` method, and I think we're going to need to here to get adequate performance on a large graph.

If you split out a "generate and collect sources" loop before a "inject synthetic targets" loop, you could make one call to `capture_snapshots` for all targets in between the two loops.

Reply:
Done